src/dist/dist.rs

```diff
@@ -824,133 +817,126 @@ fn update_from_dist_(
-                let try_next = date_from_manifest_date(toolchain_date)
-                    .unwrap_or_else(|| panic!("Malformed manifest date: {toolchain_date:?}"))
-                    .pred_opt()
-                    .unwrap();
-
-                if try_next < last_manifest {
-                    // Wouldn't be an update if we go further back than the user's current nightly.
-                    if let Some(e) = first_err {
-                        break Err(e);
-                    } else {
-                        // In this case, all newer nightlies are missing, which means there are no
-                        // updates, so the user is already at the latest nightly.
-                        break Ok(None);
-                    }
-                }
+            Some(DistError::MissingReleaseForToolchain(..)) => {
+                // no need to even print anything for missing nightlies,
+                // since we don't really "skip" them
+            }
+            _ => {
+                // All other errors break the loop
+                break Err(e);
+            }
+        };
 
-                toolchain.date = Some(try_next.format("%Y-%m-%d").to_string());
+        if let Some(backtrack_limit) = backtrack_limit {
+            if backtrack_limit < 1 {
+                // This unwrap is safe because we can only hit this if we've
+                // had a chance to set first_err
+                break Err(first_err.unwrap());
             }
         }
+
+        // The user asked to update their nightly, but the latest nightly does not have all
+        // the components that the user currently has installed. Let's try the previous
+        // nightlies in reverse chronological order until we find a nightly that does,
+        // starting at one date earlier than the current manifest's date.
+        let toolchain_date = toolchain.date.as_ref().unwrap_or(&fetched);
+        let try_next = date_from_manifest_date(toolchain_date)
+            .unwrap_or_else(|| panic!("Malformed manifest date: {toolchain_date:?}"))
+            .pred_opt()
+            .unwrap();
+
+        if try_next < last_manifest {
+            // Wouldn't be an update if we go further back than the user's current nightly.
+            if let Some(e) = first_err {
+                break Err(e);
+            } else {
+                // In this case, all newer nightlies are missing, which means there are no
+                // updates, so the user is already at the latest nightly.
+                break Ok(None);
+            }
+        }
+
+        toolchain.date = Some(try_next.format("%Y-%m-%d").to_string());
     }
 }
 
-fn try_update_from_dist_(
-    download: DownloadCfg<'_>,
-    update_hash: Option<&Path>,
-    toolchain: &ToolchainDesc,
-    profile: Option<Profile>,
+fn try_dist_update(
     prefix: &InstallPrefix,
-    force_update: bool,
-    components: &[&str],
-    targets: &[&str],
+    opts: &DistOptions<'_>,
```

Comment:
IMO this is not "just" using a struct to avoid passing a lot of parameters -- it's actually about passing a more semantically meaningful/higher-level collection of data rather than a bunch of bits and pieces which aren't individually as meaningful. Although I don't disagree in the general case that we don't pass around too much unused state, it appears me to me like the chosen abstraction is quite coherent/tightly coupled anyway, and note that even this caller (the one that's furthest down the stack so far) requires 7 of the 11 fields on the type.

I suppose we could split `DistOptions` to only contain those 7 fields and leave the other 4 directly in `InstallMethod::Dist`, would that appease your concerns? IMO that arguably makes the code harder to understand.

Reply:
I've scaled this back a little -- this last step did seem problematic since it needed a a `toolchain: ToolchainDesc` that could be different from the `ToolchainDesc` in `DistOptions::desc`.